### ISSUE TITLE
Installation should not be bloated by deploying unused files

### ISSUE THREAD

**Issue by bkil** [friendica/friendica#10070]
Installing Friendica takes too much space on small shared hosting and uploading all of this can also be time consuming. After some investigation, I've concluded that the majority of the bulk is taken up by unused files.

You can find some examples on your own instance:
* https://nerdica.net/view/asset/Chart-js/samples/charts/combo-bar-line.html
* https://nerdica.net/vendor/patrickschur/language-detection/resources/hu/hu.txt
* https://nerdica.net/view/fonts/shashape.jpg
* https://nerdica.net/vendor/michelf/php-markdown/test/resources/markdown.mdtest/Markdown%20Documentation%20-%20Syntax.text
* https://nerdica.net/spec/zot.txt
* https://nerdica.net/view/asset/Chart-js/docs/style.css
* https://nerdica.net/view/asset/Chart-js/test/fixtures/controller.line/clip/specified.json
* https://nerdica.net/view/asset/vue/yarn.lock
* https://nerdica.net/view/asset/es-jquery-sortable/CHANGELOG
* https://nerdica.net/view/asset/es-jquery-sortable/VERSION
* https://nerdica.net/images/icons/Makefile
* https://nerdica.net/tests/datasets/curl/about.head
* https://nerdica.net/view/lang/hu/messages.po
* https://nerdica.net/vendor/paragonie/certainty/test/static/data-invalid/cacert-2017-09-20.pem
* https://nerdica.net/view/asset/es-jquery-sortable/view/asset/vue/scripts/release.sh
* https://nerdica.net/view/asset/vue/scripts/git-hooks/commit-msg
* https://nerdica.net/view/theme/smoothly/js/modernizr.custom.2.5.3.js (`*.min.*` can usually be found next to most of the unminified scripts)

Here are some of the categories of bloat:
* files only used during installation (`*.sql`)
* documentation, examples, licensing
* source code, templates, build scripts or jsons, map files, dot files
* tests and related resources
* redundantly keeping the unminified or developer versions of minified assets
* not minifying every JS and CSS
* not minifying/obfuscating/optimizing PHP
* dependent libraries duplicated due to perhaps differing versions (fonts, CSS, JS)
* all fonts contain all icons and are stored in all formats

I haven't looked at the addons, but there probably would be room for optimization there as well.

Some of these are merely hidden via `.htaccess` at most deployments, like `.git`, `*.tpl` or `*.md`, which is clearly not the way to go. I wouldn't want to dictate implementation details, but probably there should be a way (`make dist`?) to prepare a folder that can be deployed onto the host.

**Comment by MrPetovan** [comment]
Thank you for your report, I know Friendica used to claim to have a light resource footprints but it is now untenable. As such, we do not offer official support for small shared hosting anymore.

We've been successfully using Composer to track and update dependencies more frequently than they used to be. This has generated some disk space overhead because packages often include test suites, documentation and source file minified versions are generated from.

However, this increase is a tiny fraction of the disk usage of even a single-user Friendica node between the database, the file system for pictures and the cache.

Feel free to comb these packages (in `view/asset/` and `vendor/` manually for extraneous files, but this will make any update more challenging to you since both the git repository and the archive update methods will restore these files.

**Comment by bkil** [comment]
Indeed I could remove all files manually - I could even write a short script that does this after each update (and I could also send it in if anyone is interested, but I'm sure there would exist more professional workarounds for this issue).

This is not a bug report per see - it is not something that is blocking me right now, because I also have a VPS running at the moment and in the long run I am planning to run multi-TB machine as well.

With this issue, I would like to propose making a public service of enabling lighter installs for the masses. This might lure in experimenters and students better, because they usually prefer completely free solutions, at least initially.

I'd like to get feedback from the project maintainers (and management) regarding this direction, because I have a few other ideas in similar vein if this is something that is welcome around here.

**Comment by MrPetovan** [comment]
Thank you for the suggestion, I personally do not think it is a viable avenue for Friendica since it isn't a tool meant to be built upon. It is a monolithic, full-featured decentralized social media server and client, which means that there isn't much to experiment with except the user experience, and this is provided completely free at one of the several public Friendica nodes.

Hosting Friendica in the current state is a somewhat costly endeavor, not only in hosting fees, but in server management knowledge and experience as well. Command-line usage is required to setup a background worker that will use significant resources to probe the network and deliver posts, potentially to hundreds of remote servers. This makes Friendica unsuitable for shared hosting environments that do not expect such a heavy application to run on their shared machines.

It seems to be possible to run Friendica on specific shared hosting environments, but it isn't a guarantee and the optimization costs to reach that guarantee are unknown right now, which makes it out of range of our two-and-change developers team.

**Comment by annando** [comment]
Concerning the needed space: Especially in this version we did a lot to reduce the database size and to improve time consuming database queries. This is something that small installations on systems with low resources can benefit from as well.

But the system still will need a lot of disk space, which is unavoidable due to its purpose.

**Comment by bkil** [comment]
Disk space taken by users themselves wouldn't always be that bad.

Consider the case when users store their media on other services (like NextCloud, PixelFed, Lufi or Lutim) and only use Friendica for texting. That's also with  assuming that they aren't following too many accounts which regularly share animated gifs or that the caches are being taken care of by cleanup scripts.

I somehow feel it to be unjust to mandate >1GB for such an instance serving only a few users - it's barely more than email.

**Comment by MrPetovan** [comment]
My current single-user node uses 2.1G for files (code + file cache) and 14G more for the database. Decentralization requires a lot of data to be copied on every single node to function smoothly. Friendica just can't be that small.

**Comment by bkil** [comment]
That sounds interesting. Which tables are the largest in your database? I guess you're storing media there. Surely indexes can also consume quite a bit.

**Comment by MrPetovan** [comment]
Here's the breakdown:

Table | Size (MB)
------|-------
photo | 2923.60
item-content | 1559.94
item | 1537.34
post-content | 825.13
conversation | 642.59
term | 579.69
contact | 460.44
item-uri | 449.81
apcontact | 273.22
post-user | 234.88
thread | 226.03
diaspora-interaction | 116.34
gcontact | 96.47
post-tag | 80.22
post | 74.67
post-thread-user | 45.17
item-activity | 42.67
fcontact | 39.59
notify | 23.48
gserver | 21.50
inbox-status | 17.61
sign | 16.70
post-thread | 15.86
tag | 14.55
participation | 12.08
cache | 12.08
post-media | 10.31
workerqueue | 9.25
contact-relation | 7.92
oembed | 7.23
user-item | 6.41
post-user-notification | 3.94
post-delivery-data | 3.31
item-delivery-data | 2.52
glink | 2.25
mail | 1.95
gserver-tag | 1.63
event | 1.63
parsed_url | 1.63
user | 0.20
conv | 0.13
spam | 0.08
item_id | 0.08
notify-threads | 0.08
session | 0.08
search | 0.06
queue | 0.06
pconfig | 0.06
profile_field | 0.06
config | 0.06
post-category | 0.05
hook | 0.05
profile | 0.05
tokens | 0.05
manage | 0.05
profile_check | 0.05
fsuggest | 0.05
gcign | 0.05
group_member | 0.05
intro | 0.05
push_subscriber | 0.05
addon | 0.05
attach | 0.03
fserver | 0.03
challenge | 0.03
locks | 0.03
2fa_trusted_browser | 0.03
verb | 0.03
advancedcontentfilter_rules | 0.03
openwebauth-token | 0.03
userd | 0.03
auth_codes | 0.03
mailacct | 0.03
poll_result | 0.03
gfollower | 0.03
group | 0.03
clients | 0.03
poll | 0.03
process | 0.03
host | 0.03
register | 0.03
2fa_app_specific_password | 0.03
delayed-post | 0.03
deliverq | 0.03
user-contact | 0.03
permissionset | 0.03
2fa_recovery_codes | 0.02
ffinder | 0.02
storage | 0.02
worker-ipc | 0.00

The `photo` table hosts picture binary data as you guessed, but even if it was hosted on an external service, the database would still be 11G big.

**Comment by bkil** [comment]
Thank you for these valuable insights. Weren't some of the tables already deprecated earlier and are just some leftover cruft after a previous update?
- `item-content`, `item-activity` https://github.com/friendica/friendica/pull/9890
- `item` https://github.com/friendica/friendica/pull/9945
- `term` https://github.com/friendica/friendica/pull/8634

Also, the number of posts/threads/conversation/etc also seems greatly excessive for a single-user instance. Is DB cleanup enabled? For a small instance, enabling `dbclean`, decreasing `dbclean_expire_conversation` to `1` and `dbclean_expire_days` to a few days might help.

It would also be nice to see row numbers as well.

Does the `contact` and `apcontact` track the whole fediverse, or why are they so huge? I tried to check the code that handles `(ap|)contact`, but am I seeing it correctly in `src/Model/(AP|)Contact.php` that such a table is almost always only extended (probably transitively), but never cleaned up from? It could very well be this very issue: #10021

**Comment by MrPetovan** [comment]
Thank you for the suggestions, they would be able to help someone having disk space issues, although it isn't my case. And setting `dbclean_expire_days` to just a few days seems handicapping to me. So I guess we're back to my main point, yes you can somewhat alleviate the disk space footprint of Friendica through careful culling, setting up an external media storage and extreme db cleaning, but this isn't the main use case I'd like us to focus on.

**Comment by bkil** [comment]
I found these interesting hacking attempts in the logs of my server:

`POST /vendor/phpunit/phpunit/src/Util/PHP/eval-stdin.php HTTP/1.1`

It is harmless as Friendica does not have this module (and these requests were blocked anyhow). However, it still underlines the fact that it is not a good idea to just blindly ship anything not generated explicitly under `dist/` to production.

Just for the kicks, I made a quick and dirty hand-crafted grep-based `make dist` filter to see what we would really need to ship and that boils down to about 30MB (out of the 150MB without `.git`) consuming a little over one third the amount of i-nodes as well. This comes down to about 10MB zipped (phar'ed). The difference might not seem much from a self-hosting perspective, but it greatly accelerates FTP'ing updates on shared hosts.

I also found a bunch of instances of using unminified resources (sometimes when the minified version was just next to the referenced one), maybe I'll find the time to submit a PR for that sometime (though, the savings in data and UX latency aren't that grave in the grand scheme of things).
